dandiapi/api/services/garbage_collection/__init__.py

```diff
@@ -0,0 +1,121 @@
+from __future__ import annotations
+
+from concurrent.futures import Future, ThreadPoolExecutor, wait
+from datetime import timedelta
+import json
+
+from celery.utils.log import get_task_logger
+from django.core import serializers
+from django.db import transaction
+from django.utils import timezone
+from more_itertools import chunked
+
+from dandiapi.api.models import (
+    AssetBlob,
+    GarbageCollectionEvent,
+    GarbageCollectionEventRecord,
+    Upload,
+)
+from dandiapi.api.storage import DandiMultipartMixin
+
+logger = get_task_logger(__name__)
+
+UPLOAD_EXPIRATION_TIME = DandiMultipartMixin._url_expiration  # noqa: SLF001
+ASSET_BLOB_EXPIRATION_TIME = timedelta(days=7)
+
+GARBAGE_COLLECTION_EVENT_CHUNK_SIZE = 1000
+
+# How long to keep GarbageCollectionEvent records around for after the garbage collection
+# is performed to allow for restoration of deleted records.
+# This should be equal to the time set in the "trailing delete" lifecycle rule.
+RESTORATION_WINDOW = timedelta(
+    days=30
+)  # TODO: pick this up from env var set by Terraform to ensure consistency?
+
+
+def _garbage_collect_uploads() -> int:
+    qs = Upload.objects.filter(
+        created__lt=timezone.now() - UPLOAD_EXPIRATION_TIME,
+    )
+    if not qs.exists():
+        return 0
+
+    deleted_records = 0
+    futures: list[Future] = []
+
+    with transaction.atomic(), ThreadPoolExecutor() as executor:
+        event = GarbageCollectionEvent.objects.create(type=Upload.__name__)
+        for uploads_chunk in chunked(qs.iterator(), GARBAGE_COLLECTION_EVENT_CHUNK_SIZE):
+            GarbageCollectionEventRecord.objects.bulk_create(
+                GarbageCollectionEventRecord(
+                    event=event, record=json.loads(serializers.serialize('json', [u]))[0]
+                )
+                for u in uploads_chunk
+            )
+
+            # Delete the blobs from S3
+            futures.append(
+                executor.submit(
+                    lambda chunk: [u.blob.delete(save=False) for u in chunk],
+                    uploads_chunk,
+                )
```

Comment:
Should we use `transaction.on_commit` here, in  case an error occurs and this transaction gets rolled back? I'm not sure if that plays nicely with waiting on those futures, you might have to define those two context managers separately.

Reply:
My thinking was that we would want the DB delete to get rolled back if an S3 delete fails. If that were to happen, we could potentially have asset blobs in the DB that have no equivalent S3 blob, which isn't a big deal and the DB asset blobs would just get cleaned up on the next garbage collection run. 
But if we delete the DB asset blobs, commit the transaction, and _then_ delete the S3 blobs, any S3 blobs that fail to delete would be "dangling" blobs with no DB record, and wouldn't get cleaned up.